sys/include/net/ng_netreg.h

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    ng_netreg  Network protocol registry
+ * @ingroup     net
+ * @{
+ *
+ * @file
+ * @brief   Definitions to register network protocol PIDs to use with
+ *          @ref ng_netapi.
+ *
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+#ifndef NETREG_H_
+#define NETREG_H_
+
+#include <inttypes.h>
+
+#include "kernel_types.h"
+#include "net/ng_nettype.h"
+
+/**
+ * @brief Number of maximum entries to @ref ng_netreg
+ * @def NG_NETREG_SIZE
+ */
+#ifndef NG_NETREG_SIZE
+#define NG_NETREG_SIZE (NG_NETTYPE_NUMOF)
+#endif
+
+/**
+ * @brief   Entry to the @ref ng_netreg
+ * @extends clist_node_t
+ */
+typedef struct ng_netreg_entry {
+    /**
+     * @see clist_node_t::next
+     */
+    struct ng_netreg_entry *next;
+    /**
+     * @see clist_node_t::prev
+     */
+    struct ng_netreg_entry *prev;
+    kernel_pid_t proto_pid;     /**< The PID of the network protocol implementation */
+
+    /**
+     * @brief   The demultiplexing context for the registering thread.
+     *
+     * @details This can be defined by the network protocol themselves.
+     *          E. g. protocol numbers / next header numbers in IPv4/IPv6,
+     *          ports in UDP/TCP, or similar.
+     */
+    uint32_t demux_ctx;
+    kernel_pid_t target_pid;    /**< The PID of the registering thread */
```

Comment:
To identify which layer you mean if you have to threads of the same implementation running (e. g. 6LoWPAN threads on top of 2  IEEE 802.15.4 radios).